src/mono/mono/metadata/class-init.c

```diff
@@ -746,65 +766,75 @@ mono_generic_class_setup_parent (MonoClass *klass, MonoClass *gtd)
 	mono_loader_unlock ();
 }
 
-struct FoundAttrUD {
-	/* inputs */
-	const char *nspace;
-	const char *name;
-	gboolean in_corlib;
-	/* output */
-	gboolean has_attr;
-};
-
 static gboolean
-has_wellknown_attribute_func (MonoImage *image, guint32 typeref_scope_token, const char *nspace, const char *name, guint32 method_token, gpointer user_data)
+has_wellknown_attribute_func (MonoImage *image, guint32 typeref_scope_token, const char *nspace, const char *name, guint32 method_token, guint32 *cols, gpointer user_data)
 {
-	struct FoundAttrUD *has_attr = (struct FoundAttrUD *)user_data;
-	if (!strcmp (name, has_attr->name) && !strcmp (nspace, has_attr->nspace)) {
-		has_attr->has_attr = TRUE;
-		return TRUE;
+	struct FoundAttrUD *attr = (struct FoundAttrUD *)user_data;
+	if (!strcmp (name, attr->name) && !strcmp (nspace, attr->nspace)) {
+		attr->has_attr = TRUE;
+		if (attr->has_value) {
+			MonoError error;
+			MonoMethod *ctor = mono_get_method_checked (image, method_token, NULL, NULL, &error);
+			if (ctor) {
+				const char *data = mono_metadata_blob_heap (image, cols [MONO_CUSTOM_ATTR_VALUE]);
+				uint32_t data_size = mono_metadata_decode_value (data, &data);
+				MonoDecodeCustomAttr *decoded_attr = mono_reflection_create_custom_attr_data_args_noalloc (image, ctor, (guchar*)data, data_size, &error);
+				mono_error_assert_ok (&error);
+				attr->value = decoded_attr->typed_args[0]->value.primitive;
+			} else {
+				g_warning ("Can't find custom attr constructor image: %s mtoken: 0x%08x due to: %s", image->name, method_token, mono_error_get_message (&error));
+			}
```

Comment:
I would prefer to pull this out into a callback for InlineArrayAttribute specifically.

I think this also needs some validation: the ctor has 1 argument, the argument really is a primitive, there are no named properties, stuff like that.
otherwise this might be leaking decoded_attr fields.

Actually do you need to free `decoded_attr`?

Reply:
> I would prefer to pull this out into a callback for InlineArrayAttribute specifically.

Do you mean into `class_has_inlinearray_attribute` or to create a new callback for inlinearray attribute?

> I think this also needs some validation: the ctor has 1 argument, the argument really is a primitive, there are no named properties, stuff like that. otherwise this might be leaking decoded_attr fields.

Good point, will be added.

> Actually do you need to free `decoded_attr`?

Probably yes.